python_bindings/src/PyGenerator.cpp

```diff
@@ -0,0 +1,193 @@
+        generator_.attr("_bind_input")(v);
+    }
+
+    void bind_input(const std::string &name, const std::vector<Func> &v) override {
+        generator_.attr("_bind_input")(v);
+    }
+
+    void bind_input(const std::string &name, const std::vector<Expr> &v) override {
+        generator_.attr("_bind_input")(v);
+    }
+
+    bool emit_cpp_stub(const std::string & /*stub_file_path*/) override {
+        // Python Generators don't support this (and *never* will, so don't ask),
+        // but don't throw an error, just return false.
+        return false;
+    }
+};
+
+class PyGeneratorFactoryProvider : public GeneratorFactoryProvider {
+public:
+    PyGeneratorFactoryProvider() = default;
+
+    std::vector<std::string> enumerate() const override {
+        py::object f = py::module_::import("halide").attr("_get_python_generator_names");
+        return args_to_vector<std::string>(f());
+    }
+    AbstractGeneratorPtr create(const std::string &name,
+                                const Halide::GeneratorContext &context) const override {
+        auto g = std::make_unique<PyGeneratorBase>(context, name);
+        if (!g->is_valid()) {
+            return nullptr;
+        }
+        return g;
+    }
+};
+
+}  // namespace
+
+void define_generator(py::module &m) {
+    auto arginfo_class =
+        py::class_<ArgInfo>(m, "ArgInfo")
+            .def(py::init<>())
+            .def(py::init([](const std::string &name, ArgInfoDirection dir, ArgInfoKind kind, std::vector<Type> types, int dimensions) -> ArgInfo {
+                     return ArgInfo{name, dir, kind, std::move(types), dimensions};
+                 }),
+                 py::arg("name"), py::arg("dir"), py::arg("kind"), py::arg("types"), py::arg("dimensions"))
+            .def_readonly("name", &ArgInfo::name)
+            .def_readonly("dir", &ArgInfo::dir)
+            .def_readonly("kind", &ArgInfo::kind)
+            .def_readonly("types", &ArgInfo::types)
+            .def_readonly("dimensions", &ArgInfo::dimensions);
+
+    auto generatorcontext_class =
+        py::class_<GeneratorContext>(m, "GeneratorContext")
+            .def(py::init<const Target &, bool, const MachineParams &>(),
+                 py::arg("target"), py::arg("auto_schedule") = false, py::arg("machine_params") = MachineParams::generic())
+            .def("target", &GeneratorContext::target)
+            .def("auto_schedule", &GeneratorContext::auto_schedule)
+            .def("machine_params", &GeneratorContext::machine_params)
+            // TODO: handle get_externs_map() someday?
+            .def("__repr__", [](const GeneratorContext &context) -> std::string {
+                std::ostringstream o;
+                o << "<halide.GeneratorContext " << context.target() << ">";
+                return o.str();
+            });
+
+    py::object scope = m.attr("__dict__");
+    py::exec(py::str((const char *)builtin_helpers_src, builtin_helpers_src_length), scope);
```

Comment:
After #6815 and the `src/` rearrangement, I don't think we'll need or want to do this binary2cpp / py::exec dance.

Reply:
Cool, I eagerly await :-)